src/frame_buffer/single_frame_allocator.rs

```diff
@@ -63,49 +51,44 @@ impl SingleFrameAllocator {
     ///   let allocator = SingleFrameAllocator::new();
     /// # }
     /// ```
-    pub fn new() -> Self {
-        let frame_allocator = Self {
-            is_allocated: false,
-        };
-
-        crit_sect::locked(|cs| {
-            let prev_frame_allocator = FRAME_ALLOCATOR.borrow(cs).replace(Some(frame_allocator));
-            assert!(prev_frame_allocator.is_none());
-        });
-
-        let phantom_allocator = Self { is_allocated: true };
-        phantom_allocator
+    pub const fn new() -> Self {
+        Self {
+            is_allocated: AtomicBool::new(false),
+            data: UnsafeCell::new([0; Self::BUF_SIZE]),
+        }
     }
 
     /// Releases an allocated frame
     ///
     /// Function called when any [`FrameBuffer`](super::frame_buffer::FrameBuffer) created by this
     /// allocator is dropped.
-    fn release_frame(_frame: &mut FrameBuffer, _metadata: DropMetadata) {
-        SingleFrameAllocator::use_frame_allocator(|fa| {
-            assert!(fa.is_allocated);
-            fa.is_allocated = false;
-        });
+    pub fn release_frame(&self, _frame: &mut FrameBuffer) {
+        self.is_allocated
```

Comment:
There shouldn't be a race in the release procedure. I would `assert` (or at least `debug_assert`) that the frame is allocated at the moment.

Reply:
If I got ordering correctly this shouldn't be possible. But `debug_assert` is always fine.